### PR TITLE
Constexpr

### PR DESCRIPTION
**Pull request by mrdudz**
This should fix issue #1538, #1613 and similar related problems. Not heavily tested yet (if you can think of a particular edge case that needs testing, please tell)

### CODE REVIEW

**Comment by oliverschmidt** [comment]
I presume it was an accident to ask me for a review. I'm only working on Apple II related aspects of cc65.

File: src/ca65/studyexpr.c [hunk 1]
```diff
@@ -223,19 +224,33 @@ static void ED_MergeAddrSizeAND (ExprNode* Expr, ExprDesc* ED, const ExprDesc* R
...
+        ConstSize = ADDR_SIZE_FAR;
     } else {
-        ED->AddrSize = ADDR_SIZE_DEFAULT;
+        ConstSize = ADDR_SIZE_DEFAULT;
+    }
+
+    if (Size == ADDR_SIZE_DEFAULT) {
+        ED->AddrSize = ConstSize;
+    } else if (ConstSize == ADDR_SIZE_DEFAULT) {
+        ED->AddrSize = Size;
```
[the remark this answers is not in the document]
Comment (mrdudz) [reply src/ca65/studyexpr.c:250]: Yes, in that case the size is unknown and that means ADDR_SIZE_DEFAULT :)

A testcase for this situation would probably not hurt either.... mmh

[the remark this answers is not in the document]
Comment (spiro-trikaliotis) [reply src/ca65/studyexpr.c:250]: Ok, I was not sure if it was intentionally or by accident.
I must admit that this code surprises me every time I look at it, so I am very careful when trying to change something.

[the remark this answers is not in the document]
Comment (mrdudz) [reply src/ca65/studyexpr.c]: It actually wasn't entirely correct, see my fix - that makes more sense :) The constant part actually can never have an unknown size, and we can assume that if it its neither 8,16 or 24bit, then it is 32bit.

Still trying hard to come up with proper testcases for the the edge cases of this thing

[the remark this answers is not in the document]
Comment (spiro-trikaliotis) [reply src/ca65/studyexpr.c]: Ok, good to know that my comment did make sense. ;)

Do you already have some idea how some test cases could be written? We might make some progress by discussing them together.

[the remark this answers is not in the document]
Comment (mrdudz) [reply src/ca65/studyexpr.c]: ehrm, did you see the actual test program in the PR? :) all tests that check valid code can be put into this one. the cases that are invalid code and should cause the assembler to error out will require extra tests (but those are probably not as critical - the important ones are the edge cases that should be valid, and check if they produce the correct result)

**Comment by mrdudz** [comment]
If noone objects, i am going to merge this soon - it seems safe enough and if it really breaks something should be easy to fix

Just for the records, this is how i tried finding edge cases (but wasnt very successful for the interesting ones). If you can make it show any ADDR_SIZE_DEFAULT (value = 0) in this function, please tell how you did it, so we can produce a test case for that situation :)

```
diff --git a/src/ca65/studyexpr.c b/src/ca65/studyexpr.c
index 40e9b7aab..5d777f176 100644
--- a/src/ca65/studyexpr.c
+++ b/src/ca65/studyexpr.c
@@ -194,6 +194,7 @@ static void ED_MergeAddrSizeAND (ExprNode* Expr, ExprDesc* ED, const ExprDesc* R
     int ConstL, ConstR;
     int Size, ConstSize;
     long Val, ValR;
+printf("ED_MergeAddrSizeAND AddrSizeL:%d AddrSizeR:%d\n", ED->AddrSize, Right->AddrSize);
 
     if (ED->AddrSize == ADDR_SIZE_DEFAULT) {
         /* If ED is valid, ADDR_SIZE_DEFAULT gets always overridden, otherwise
@@ -220,10 +221,12 @@ static void ED_MergeAddrSizeAND (ExprNode* Expr, ExprDesc* ED, const ExprDesc* R
     /* Check if either side of the expression is constant */
     ConstL = IsConstExpr (Expr->Left, &Val);
     ConstR = IsConstExpr (Expr->Right, &ValR);
+printf("ConstL:%d ConstR:%d\n", ConstL, ConstR);
     if (!ConstL && !ConstR) {
         /* If neither part of the expression is constant, the above is all we can do */
         return;
     }
+printf("AddrSizeL:%d AddrSizeR:%d\n", ED->AddrSize, Right->AddrSize);
     /* We start assuming the left side is constant, left value is in Val, right
        size in Size */
     Size = Right->AddrSize;
@@ -243,6 +246,7 @@ static void ED_MergeAddrSizeAND (ExprNode* Expr, ExprDesc* ED, const ExprDesc* R
     } else {
         ConstSize = ADDR_SIZE_LONG;
     }
+printf("Size:%d ConstSize:%d\n", Size, ConstSize);
 
     if (Size == ADDR_SIZE_DEFAULT) {
         ED->AddrSize = ConstSize;
@@ -250,6 +254,7 @@ static void ED_MergeAddrSizeAND (ExprNode* Expr, ExprDesc* ED, const ExprDesc* R
         /* use the smaller of the two sizes */
         ED->AddrSize = (ConstSize < Size) ? ConstSize : Size;
     }
+printf("AddrSize:%d\n", ED->AddrSize);
 }
```
